### ISSUE TITLE
Random sampling while reading tfrecord

### ISSUE THREAD

**Issue by dorucia** [NVIDIA/DALI#1835]
TL;DR
I have 1000x3 float data in tfrecord and I want to read it as 500x3 data where 500 is randomly sampled from 1000. I want it to be random every time I read it.

Hi, I'm using PyTorch + DALI + tfrecord and GPUs.

I stored 3D point cloud (xyz) from several frames in the tfrecord file.
The number of points in each frame is different.
For example, at frame 1, I have 1000 points so the information is stored as 1000x3 in float type.
At frame 2, I have 2000 points so it is 2000x3 data.
Therefore, I use 
`"cloud" :     tfrec.VarLenFeature(tfrec.float32, 0.)`
to load the point cloud in DALI's `ops.TFRecordReader`.

To get this data into PyTorch's tensor, I first pad constant values to match the maximum number of points in the batch.
In the above example, the second frame has the biggest number of points (2000).
Therefore, I pad the first point cloud with constant 1000x3 values.
A batch in PyTorch tensor with batch_size=2 will look like this: 2 x 2000 x 3.

For each iteration, I want to randomly sample 500 points at each frame to make 2 x 500 x 3 as an input to the network.
In order to do that, I should remember the number of points at each frame.
For example, for frame 1, although the batch is loaded as 2000x3 shape, only first 1000x3 are real points and the rest of 1000x3 are padded constant.
Therefore, I sample 500 points from 0~999 range.

This approach works, however, it seems it has an overhead.
The GPU usage is around 60% on the nvidia-smi.

I think it would be great if we can directly do random sampling when we load the data from tfrecord.
I checked two functions in DALI.
One is Slice and the other is ElementExtract.
For the Slice operator, it randomly gets a chunk. It does not look like a random access that I described above.
For ElementExtract, I'm not sure whether it can achieve my goal.

How can I do this?

**Comment by JanuszL** [comment]
Hi,
As I understand correctly you have an entry with some data in TFRecord, and you want to randomly sample only part of that entry. I don't think that there is any build-in operator in DALI that can be applied here.
What you can do is to extend [DALI on your own](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/custom_operations/index.html) - the first start with [python operator](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/custom_operations/python_operator.html) to see if you can achieve what you want and then try with the [native one](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/custom_operations/custom_operator/create_a_custom_operator.html).

**Comment by dorucia** [comment]
Thanks for the reply.
I'll try those custom operators.

I think random sampling is a common operation to deal with 3D data based on point clouds.
It would be great if this feature canl be supported by DALI in the future.

**Comment by JanuszL** [comment]
@dorucia - I think DALI is far from supporting 3D data based on point clouds yet. Sampling is one thing, but I would think we would focus on some transformations on it first.

**Comment by dorucia** [comment]
I made a simple python operator as you suggested and it works like a charm.
Thanks.

I have a related question.

Is it better to use LMDB rather than tfrecord when I use PyTorch+DALI  in terms of data loading speed?

**Comment by JanuszL** [comment]
> Is it better to use LMDB rather than tfrecord when I use PyTorch+DALI in terms of data loading speed?

We didn't do any speed comparison between containers. In both cases, you have protobuf entry stored which you need to parse. The only advantage LMDB may have is that you don't need to create a separate index file for DALI, on the other hand, it is some sort of DB so it can be slower than just a plain file like TFRecord.

**Comment by dorucia** [comment]
It seems random shuffling in the python operator always gives the same result.
I tested both numpy and cupy.
Here is a code.

```
def sample_points(cloud):
    cloud_sampled = cupy.copy(cloud)
    c_mask = cupy.zeros(cloud.shape[0], dtype=int)
    c_mask[:512] = 1
    cupy.random.shuffle(c_mask)
    cloud_sampled = cloud_sampled[c_mask.nonzero()]
    return cloud_sampled


class TFRecordPipeline(Pipeline):
    def __init__(self, tfrecord, tfrecord_idx, batch_size, num_threads, device_id, shard_id, num_shards, shuffle):
        super(TFRecordPipeline, self).__init__(batch_size, num_threads, device_id, seed=12+device_id, exec_async=False, exec_pipelined=False)

        self.input = ops.TFRecordReader(path = tfrecord,
                                        index_path = tfrecord_idx,
                                        features = {"cloud" : tfrec.VarLenFeature(tfrec.float32, 0.)},
                                        shard_id = shard_id, num_shards = num_shards, random_shuffle = shuffle)
        self.reshape = ops.Reshape(shape=[-1,3], device='gpu')  # number_of_points_at_the_frame x xyz
        self.sample_points = ops.PythonFunction(function=sample_points, num_outputs=1, device='gpu')

    def define_graph(self):
        inputs = self.input()
        cloud = self.sample_points(self.reshape(inputs['cloud'].gpu()))

        return cloud
```

Then, I define

```
data_loader = TFRecordPipeline(tfrecord, tfrecord_idx,
                               batch_size=opt.batch_size, num_threads=opt.num_workers, device_id=opt.local_rank,
                               shard_id=opt.local_rank, num_shards=opt.num_gpus, shuffle=shuffle)
data_loader.build()

data_iter = DALIGenericIterator(data_loader, ['cloud'], num_data)
```

Then, I print the loaded data like this

```
t1 = next(train_iter)[0]
t2 = next(train_iter)[0]
print('t1: ', t1['cloud'], 't2: ', t2['cloud'])
```

It seems t1 and t2 have the same point cloud which means the cupy.random.shuffle() gives the same result.

I tried two things to fix this
1. Reset random seed by adding
```
cupy.random.seed()  <- new line in the sample_points function
cupy.random.shuffle(c_mask)
```

2. Set seed of the ops.PythonFunction:
```
ops.PythonFunction(function=sample_points, num_outputs=1, device='gpu', seed=int(time.time()))
```

but neither of them fix the issue.

Do you have an idea?

**Comment by JanuszL** [comment]
@banasraf - any idea?

**Comment by JanuszL** [comment]
@dorucia - you can add some debug prints inside `sample_points` to dump input and output to see what is happening there.

**Comment by dorucia** [comment]
oh, i found a bug from my side.
it seems working

**Comment by dorucia** [comment]
I took a look at some c++, cuda operators in DALI repo.
It seems it is difficult for me to efficiently implement random sampling feature in c++ and cuda.
So I will use Python operator for now.

I think the random sampling is not restricted to point clouds.
For example, we might need to sample random video frames or sample from feature map at random, etc.
It would be great if anyone interested in this topic jumps in.
